include/verilated_trace_imp.cpp

```diff
@@ -456,7 +464,83 @@ void VerilatedTrace<VL_DERIVED_T>::dumpvars(int level, const std::string& hier)
     }
 }
 
-template <> void VerilatedTrace<VL_DERIVED_T>::dump(uint64_t timeui) VL_MT_SAFE_EXCLUDES(m_mutex) {
+#ifdef VL_TRACE_PARALLEL
+template <>  //
+void VerilatedTrace<VL_SUB_T, VL_BUF_T>::parallelWorkerTask(void* datap, bool) {
+    ParallelWorkerData* const wdp = reinterpret_cast<ParallelWorkerData*>(datap);
+    wdp->m_cb(wdp->m_userp, wdp->m_bufp);
+    wdp->m_done.store(true);
+}
+
+template <>
+VL_ATTR_NOINLINE void VerilatedTrace<VL_SUB_T, VL_BUF_T>::waitForFlag(std::atomic<bool>& flag) {
+    // Spin waiting for flag to become true. This is a separte function so it
+    // shows up clearly when porfiling
+    while (!flag.load()) VL_CPU_RELAX();
```

Comment:
Uhm, but we do exactly this everywhere else:

https://github.com/verilator/verilator/blob/0722f47539113382db67d94f46e28cb8b22c2a37/include/verilated_threads.h#L180-L185

and

https://github.com/verilator/verilator/blob/0722f47539113382db67d94f46e28cb8b22c2a37/include/verilated_threads.h#L121-L128

In those we do VL_LOCK_SPINS, relaxing each time around the loop, and we exhausted VL_LOCK_SPINS, then we suspend the thread on a condition variable. All the other spin-lock implementations I seen suggest doing what we do already. Not sure what you intended here. AFAIK the pause/yield instructions is just supposed to hint the hardware scheduler to prioritise other threads on a multi-threaded core.

Reply:
Sorry, I was confusing yield and relax. We need to do a loop with relax, then after a number of tries do a yield.  Otherwise the yield will only happen after the OS forces a timeslice switch which is very long. If someone did not use numactl or the system is overloaded (like running test regress) it may make the runtime go from horrible (1000x) to impossible (100000x+).   It's possible we are missing consistent relax loop/yield some places, that should be fixed.

Reply:
Ok, I will go add a "yield thread" after fixed spins, as otherwise we just spin all the way here.